source/grid/tria_accessor.cc

```diff
@@ -1869,6 +1869,144 @@ CellAccessor<dim, spacedim>::neighbor_of_coarser_neighbor (const unsigned int ne
 
 
 template <int dim, int spacedim>
+bool
+CellAccessor<dim, spacedim>::has_periodic_neighbor (const unsigned int i_face) const
+{
+  typedef TriaIterator<CellAccessor<dim, spacedim>> cell_iterator;
+  /* my_it : is the iterator to the current cell. */
+  cell_iterator my_it(*this);
+  if (this->tria->periodic_face_map.find(std::pair<cell_iterator, unsigned int>(my_it, i_face))
+      != this->tria->periodic_face_map.end())
+    return true;
+  return false;
+}
+
+
+
+template <int dim, int spacedim>
+TriaIterator<CellAccessor<dim,spacedim> >
+CellAccessor<dim, spacedim>::
+neighbor_or_periodic_neighbor (const unsigned int i_face) const
+{
+  typedef TriaIterator<CellAccessor<dim, spacedim>> cell_iterator;
+  typedef std::pair<cell_iterator, unsigned int> cell_face_pair;
+  /* my_it        : the iterator to the current cell.
+   * my_face_pair : the pair reported by periodic_face_map as its first pair being
+   *                the current cell_face.
+   */
+  cell_iterator my_it(*this);
+  typename std::map<const cell_face_pair, cell_face_pair>::const_iterator my_face_pair =
+    this->tria->periodic_face_map.find(std::pair<cell_iterator, unsigned int>(my_it, i_face));
+  if (my_face_pair != this->tria->periodic_face_map.end())
+  {
+    /*
+     * Some assertions might be required !
+     */
+    return my_face_pair->second.first;
+  }
+  else
+    return neighbor(i_face);
+}
+
```

Comment:
In most cases faces are not at the boundary. Hence, I would return the neighbor 
,if the cell is not at the boundary, and only search for the periodic neighbor otherwise.